controllers/register_test.go

```diff
@@ -0,0 +1,141 @@
+package controllers
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/gutakk/go-google-scraper/db"
+	"github.com/gutakk/go-google-scraper/models"
+	"github.com/gutakk/go-google-scraper/tests"
+	"github.com/stretchr/testify/suite"
+	"gopkg.in/go-playground/assert.v1"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+func TestDisplayRegister(t *testing.T) {
+	engine := tests.GetRouter(true)
+	new(RegisterController).applyRoutes(engine)
+
+	response := tests.PerformRequest(engine, "GET", "/register", nil, nil)
+	p, err := ioutil.ReadAll(response.Body)
+	pageOK := err == nil && strings.Index(string(p), "<title>Register</title>") > 0
+
+	assert.Equal(t, http.StatusOK, response.Code)
+	assert.Equal(t, true, pageOK)
+}
+
+type DBTestSuite struct {
+	suite.Suite
+	engine   *gin.Engine
+	formData url.Values
+	headers  http.Header
+}
+
+func (s *DBTestSuite) SetupTest() {
+	testDB, _ := gorm.Open(postgres.Open(tests.ConstructTestDsn()), &gorm.Config{})
+	db.GetDB = func() *gorm.DB {
+		return testDB
+	}
+
+	_ = db.GetDB().AutoMigrate(&models.User{})
+
+	s.engine = tests.GetRouter(true)
+	registerController := &RegisterController{}
+	registerController.applyRoutes(s.engine)
+
+	s.headers = http.Header{}
+	s.headers.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	s.formData = url.Values{}
+	s.formData.Set("email", "test@hello.com")
+	s.formData.Set("password", "123456")
+	s.formData.Set("confirm-password", "123456")
+}
+
+func (s *DBTestSuite) TearDownTest() {
+	db.GetDB().Exec("DELETE FROM users")
+}
+
+func TestDBTestSuite(t *testing.T) {
+	suite.Run(t, new(DBTestSuite))
+}
+
+func (s *DBTestSuite) TestRegisterWithValidParameters() {
```

Comment:
Now that `saveUser` method has been moved to the model, many of those tests can be moved at the model level (right now there are not tests). So the controller could have simpler tests (1 test for valid data and 1 for invalid data), that's the advantage of moving logic to model or service objects.

Reply:
Add user model unit test in 1bfcb6ebf3caab5a5f0d2804f98549839b9655d7 295c3b970e6c47f80b1461440bd6dafe40e25892

I think we cannot remove invalid data test cases because all of the invalid data test cases are validation. IMO this is controller responsibility.

What do you think?

Reply:
Also using `faker` to generate random data for each test case

Reply:
> I think we cannot remove invalid data test cases because all of the invalid data test cases are validation. IMO this is controller responsibility.

Controllers must validate the user input indeed. That's where the struct `RegisterForm` is useful. And tests in controllers should be reserved to HTTP code checks and errors as done here. 

Right now, the controller tests seem fine but the test on the models are a bit too limited. When adding more tests to models, you will realize that some tests on the controller are somewhat redundant. But if there are as many tests on the controller as on the model then there are no issues. It's a question of efficiency: it's better to test thoroughly the business logic instead of controller flows.

Reply:
> > I think we cannot remove invalid data test cases because all of the invalid data test cases are validation. IMO this is controller responsibility.
> 
> Controllers must validate the user input indeed. That's where the struct `RegisterForm` is useful. And tests in controllers should be reserved to HTTP code checks and errors as done here.
> 
> Right now, the controller tests seem fine but the test on the models are a bit too limited. When adding more tests to models, you will realize that some tests on the controller are somewhat redundant. But if there are as many tests on the controller as on the model then there are no issues. It's a question of efficiency: it's better to test thoroughly the business logic instead of controller flows.

Update error message return type and add more user model test cases in ae0d6f9411565b4dfe1efa6e67c82417147c5873 78603a7083efc30b6932fe6c7fb0cb08b109d551